### ISSUE TITLE
Requires .NET 3.5

### ISSUE THREAD

**Issue by timothyscottsmith** [heaths/sdb2xml#4]
Hi,

This application is currently being used in another project, specifically [FU.WhyAmIBlocked](https://github.com/AdamGrossTX/FU.WhyAmIBlocked). This tool can be run on Windows 10, while also being run from remote devices, but many Windows 10 machines will not have .NET 3.5 already installed. Is there any way to make this more natively compatible with Windows 10 devices? .NET 3.5 is not installed by default on Windows 10 devices, and .NET 3.5 [has not been a default install since Windows 7](https://docs.microsoft.com/en-us/archive/blogs/astebner/mailbag-what-version-of-the-net-framework-is-included-in-what-version-of-the-os) (it's _available_ after Windows 7, but not installed by default).

It seems, by and large, that modernization of this application would include increasing its default .NET requirement to something more recent, such as .NET 4.5.1 **and** higher. This would work to support Server 2012 R2, 2016, and 2019, as well as Windows 8.1 and Windows 10. I feel that, in this way, it would penalize more recent operating systems less by requiring a newer framework.

Admittedly, I have a high level of ignorance when it comes to C#. I don't know how difficult this is, or if it's even possible, but I figured that asking couldn't hurt. I certainly appreciate your time in this regard and hope that this request isn't completely crazy.

Timothy

**Comment by heaths** [comment]
Normally, upgrading to a newer Framework (or Core) is easy. I started to do that on this project but it was decompiled from my original binary since the source was lost. It's not been a priority to fix but I am accepting PRs.

**Comment by heaths** [comment]
You could also look up app.config files for .NET and try to force it to a newer version. Hypothetically, it should work. You'd add an sdb2xml.exe.config file with `<supportedFramework>` or something like that. It's been a while since I've had to do that, but you should be able to find plenty of information online if you want to try that.

**Comment by timothyscottsmith** [comment]
I really appreciate the help on this. Through some research, I did find what I need. I appreciate your willingness to help and reply to me with some information.

For anyone that wanders through here, this is the code I used:

```
<?xml version="1.0"?>
<configuration>
  <startup>
    <supportedRuntime version="v4.0"/>
  </startup>
</configuration>
```
Name the file "sdb2xml.exe.config" and place it in the same directory as the executable. I don't receive any prompts for .NET 3.5 anymore.

Thank you again!

Timothy

**Comment by timothyscottsmith** [comment]
It appears I spoke too soon. The config file will silence the prompt to install .NET 3.5, but the application does not run. Seeing that I know nothing about C#, this will not be something I will be able to fix.

I just want to reiterate that I appreciate the time you have taken to respond to me. Take care.

Timothy

**Comment by heaths** [comment]
What was your command line? I just tried this and it worked on my machine. I do have .NET 3.5 enabled (needed for some development efforts), but Process Monitor shows .NET 4 was loaded.

Everything in .NET 3.5 (especially that this app uses) is present in .NET 4.0 and Core, so there's no reason it can't run as long as the right runtime loaded.

**Comment by MattiasC85** [comment]
Not done yet. But if you'd like to, try out this version.
https://github.com/MattiasC85/Scripts/tree/master/sdb2xmlModified
:)

**Comment by timothyscottsmith** [comment]
> Not done yet. But if you'd like to, try out this version.
> https://github.com/MattiasC85/Scripts/tree/master/sdb2xmlModified
> :)

THANK you! This appears to work, and there is full output in the XML file.

@heaths This was simply the command line I was using:

`sdb2xml.exe "C:\temp\appraiser.sdb_Expanded_ver_2469.sdb" -out C:\temp\appraiseroutput.xml`

It wouldn't throw an error at all after adding the config file, but the XML would also be empty. On an old Windows 7 machine I had (not the best test case), it would output _some_ XML, but then error out complaining about memory corruption.

**Comment by MattiasC85** [comment]
I just noticed something. Need to see if it is related.. Don't use that exe until I update this post or write a new one. 
**DO NOT USE THAT Exe <- Discovered a huge issue.**
Will upload a fixed one.

**Comment by timothyscottsmith** [comment]
Cryptic, but okay.

**Comment by MattiasC85** [comment]
In my tries, if the sdb is packed, it will create thousands of .bin files. *facepalm*

**Comment by timothyscottsmith** [comment]
Yeah that's a problem haha. Thankfully I ran it against some expanded SDB, they weren't packed, so I didn't run into an incident. Appreciate the help!

**Comment by heaths** [comment]
@MattiasC85 would you be willing to submit a PR to this instead? As the original author people have been contacting about this, it'd be great to keep the main trunk up to date. Certainly people are welcome to fork it and make modifications, but the great thing about open source is contributing back generally great changes for everyone's benefit.

**Comment by MattiasC85** [comment]
@heaths Will give you more info tomorrow. But I'm not too familiar with github and the shortenings, so what is a PR?
I'm all about sharing knowledge and improvements. When I'm sure that my modification isn't to blame for bin-file case and feeling less ashamed about of how my code looks like I will start to look into of how to fork and all that stuff :P

**Comment by heaths** [comment]
No problem! <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests> might be a good start to forking and creating PRs, but I'm happy to help on specific questions. A PR, for example, is just a request to merge your topic branch in with a target branch like `master` here in this main repo.

**Comment by MattiasC85** [comment]
I did one single edit to the original code (else it would throw an error even with .Net core 3.1 as the target).
Changed         

[DllImport(APPHELP, CharSet = CharSet.Unicode)]
internal static extern string SdbGetStringTagPtr(Pdb pdb, int tiWhich);

To

[DllImport(APPHELP, CharSet = CharSet.Unicode)]
internal static extern IntPtr SdbGetStringTagPtr(Pdb pdb, int tiWhich);

and to reflect that change also              
case TypeCode.String:
                                return Marshal.PtrToStringAuto(NativeMethods.SdbGetStringTagPtr(pdb, id));

in tag.cs.

I don't think the change is the cause, even though I'm not 100% sure, but using "-extract"  the program creates 10's of thousands bin files if run against an unpacked version of C:\Windows\System32\appraiser\appraiser.sdb.

Unpacked using the 3rd party  FU.WhyAmIBlocked so not unpacked within the code.

**Comment by MattiasC85** [comment]
I'm no programmer at all. So please be kind.  https://github.com/MattiasC85/sdb2xml

Added some stuff, like beeing able to target C:\Windows\System32\appraiser\appraiser.sdb (even if it's packed), a temporary unpacked sdb-file and a cancelhandler that should delete the file if ctrl+c is pressed.
Also added an unpacked version of appraiser.sdb if you want to do any tests.

But doing a fork and convert the whole thing to a .NET framework 4.x.x application. I feel like that's no change to the current code, but a whole other app. :P

**Comment by MattiasC85** [comment]
@timothyscottsmith I've only seen those .bin files beeing created if you use the -extract argument and that is exactly what that command should do. Don't think my changes are to blame but the appraiser db is just simply big or there's some new element to it which is causing it. 

But for a "belt and suspenders" approach, do a check once or twice just to make sure after you've done what you needed with the exe. 

Standing the in root of the systemdrive.

(Get-ChildItem *.bin -Recurse -Force -ErrorAction SilentlyContinue).Directory | Group-Object -NoElement | Where-Object -Property Count -gt 50 | Select-Object -Property Name,Count | Format-Table
![githubpicIssue](https://user-images.githubusercontent.com/44426149/115080440-086cd300-9f03-11eb-9ad0-ea4d4e83a28d.png)
